### DISCUSSION TITLE
ESPHome stable/beta/dev choice

### DISCUSSION THREAD

**Discussion by nagyrobi** [tteck/Proxmox#975] [General]
ESPHome has 3 flavours (also available as 3 different Home Assistant add-ons):
- stable
- beta
- dev

Would be nice to have a choice in the installer script to select which flavour to install. Not in the advanced settings, just at the beginning. The use case is to have the possibility to have the 3 different flavours installed as separate LXCs.

Not sure if the script also performs an update, however the `dev` flavour has daily updates (or more often when the leader developer performs an immediate build) so it would be very handy to be able to do the update through the script, at the flavour chosen during the installation.

The only question that remains to be thought about is how to mount a shared storage between them (in a Proxmox cluster environment), in order for all 3 LXCs to use the same ESPHome yaml configuration directory (just like they do as HA add-ons).

**Comment by tteck** [comment]
https://esphome.io/guides/faq.html#how-do-i-update-to-the-latest-version

After creation, pick your flavor and run in the LXC console
- `pip3 install -U esphome` for **stable**
- `pip3 install --pre -U esphome` for **beta**
- `pip3 install https://github.com/esphome/esphome/archive/dev.zip` for **dev**

Then a `systemctl restart esphomeDashboard`

**Reply by tteck** [reply]
You can update each using the same command.

**Reply by nagyrobi** [reply]
Wouldn't it be nice for the LXC script to prompt for a choice when first run?

**Reply by tteck** [reply]
With the ease of changing branches now, not worth the extra coding IMO
Most should stay on the stable branch (no confusing with options)

**Comment by nagyrobi** [comment]
Can you give me a hint on how to make a common shared directory between multiple LXC containers? Would be important to run in cluster environment, and be able to migrate.

For example, I'd host the files on the LXC running the`stable` flavour, and share it out as a network share to the other two. Any recommendation? Like NFS server on the first LXC and NFS client mounts on the other two?

**Reply by nagyrobi** [reply]
OK solved this myself. Posting here as others may also need this. 
Install ESPHome LXC as _privileged_ containers (unprivileged = no).

First install the container for the stable flavour.
1. set a static IP address for it in your network (say, 192.168.1.10)
2. Power it off, and add to the config file manually the entry `lxc.apparmor.profile: unconfined` 
(edit the config on the host with `nano /etc/pve/lxc/<CTID>.conf`)
4. start it, then then `apt install nfs-kernel-server -y`
5. edit the share to point to the esphome yaml config dir: `nano /etc/exports` 
and add `/root/config/   192.168.1.0/24(rw,sync,no_subtree_check)`
7. run `exportfs -ar`
8. `service nfs-server restart` should not complain
9. put some dummy testfile in `/root/config/` just to be able to check from the other machine

Install a new container for the beta flavour (see above post to change stable to beta)
1. in container options enable Feature NFS and reboot.
2. then `apt install nfs-common`
3. try to mount the share from command line 
`mount -t nfs -o vers=3 192.168.1.10:/root/config/ /root/config/`
5. check in `/root/config` do you see the dummy file from the other LXC?
6. if yes, you can add to `/etc/fstab`  the entry to mount at boot: 
`192.168.1.10:/root/config	/root/config	nfs	vers=3	0 0`

You can repeat the last section for the dev flavour.

Credits to [here](https://virtualizeeverything.com/2022/06/14/nfs-server-in-proxmox-ct-lxc-container/) and [here](https://theorangeone.net/posts/mount-nfs-inside-lxc/).

To install a `smb` share in order to access the yaml files like the addon used to share them, follow steps from [here](https://www.techrepublic.com/article/how-to-create-a-samba-share-on-ubuntu-server-20-04/) on the stable container.
